stdlib/public/core/RuntimeFunctionCounters.swift

```diff
@@ -0,0 +1,506 @@
+  let mirror = Mirror(
+    legacy: _reflect(value),
+    subjectType: type(of: value))
+
+  let count = mirror.children.count
+
+  let id: ObjectIdentifier?
+  let ref: UnsafeRawPointer?
+  if type(of: value) is AnyObject.Type {
+    // Object is a class (but not an ObjC-bridged struct)
+    let toAnyObject = _unsafeDowncastToAnyObject(fromAny: value)
+    ref = UnsafeRawPointer(Unmanaged.passUnretained(toAnyObject).toOpaque())
+    id = ObjectIdentifier(toAnyObject)
+  } else if type(of: value) is Builtin.BridgeObject.Type {
+    ref = UnsafeRawPointer(
+      Builtin.bridgeToRawPointer(value as! Builtin.BridgeObject))
+    id = nil
+  } else if type(of: value) is Builtin.NativeObject.Type  {
+    ref = UnsafeRawPointer(
+      Builtin.bridgeToRawPointer(value as! Builtin.NativeObject))
+    id = nil
+  } else if let metatypeInstance = value as? Any.Type {
+    // Object is a metatype
+    id = ObjectIdentifier(metatypeInstance)
+    ref = nil
+  } else {
+    id = nil
+    ref = nil
+  }
+
+  if let theId = id {
+    // Bail if this object was seen already.
+    if visitedItems[theId] != nil {
+      return
+    }
+    // Remember that this object was seen already.
+    let identifier = visitedItems.count
+    visitedItems[theId] = identifier
+  }
+
+  // If it is a reference, add it to the result.
+  if let ref = ref {
+    references.append(ref)
+  }
+
+  // Recursively visit the children of the current value.
+  var currentIndex = mirror.children.startIndex
+  for _ in 0..<count {
+    let (_, child) = mirror.children[currentIndex]
+    mirror.children.formIndex(after: &currentIndex)
+    _collectAllReferencesInsideObjectImpl(
+      child,
+      references: &references,
+      visitedItems: &visitedItems)
+  }
+}
+
+/// This is a namespace for runtime functions related to management
+/// of runtime function counters.
+public // @testable
+struct _RuntimeFunctionCounters {
+  public typealias RuntimeFunctionCountersUpdateHandler =
+   @convention(c) (_ object: UnsafeRawPointer, _ functionId: Int64) -> Void
+
+  public static let runtimeFunctionNames =
+    getRuntimeFunctionNames()
+  public static let runtimeFunctionCountersOffsets =
+    _RuntimeFunctionCounters.getRuntimeFunctionCountersOffsets()
+  public static let numRuntimeFunctionCounters =
+    _RuntimeFunctionCounters.getNumRuntimeFunctionCounters()
+  public static let runtimtFunctionNameToIndex: [String : Int] =
+    getRuntimeFunctionNameToIndex()
+
+  /// Get the names of all runtime functions whose calls are being
+  /// tracked.
+  @_silgen_name("getRuntimeFunctionNames")
+  static public func _getRuntimeFunctionNames() ->
+    UnsafePointer<UnsafePointer<CChar>>
+
+  static public func getRuntimeFunctionNames() -> [String] {
+    let runtimeFunctionNames = _RuntimeFunctionCounters._getRuntimeFunctionNames()
+    let numRuntimeFunctionCounters =
+      _RuntimeFunctionCounters.getNumRuntimeFunctionCounters()
+    var runtimtFunctionNames : [String] = []
```

Comment:
Since you already know the future length of a resulting array, calling `runtimeFunctionNames.reserveCapacity(numRuntimeFunctionCounters)` will help avoid reallocation.